Add CI workflow to build SADL and WebSADL

tuxji opened the pull request:
Set up a GitHub Actions workflow to build SADL source, build WebSADL
extension, publish WebSADL npm package, and push WebSADL image to
Docker Hub.  The workflow (which came from Typefox) has not been
tested and the authentication for the publish npm package step may
require action by Typefox.  I have created the necessary
NPM_AUTH_TOKEN, DOCKERHUB_USERNAME, and DOCKERHUB_TOKEN secrets in the
sadl GitHub settings.

It was not clear to me whether we should push the WebSADL Docker image
to the theiaide or sadl repository in Docker Hub.  I've updated some
documentation files and the workflow on the assumption we'll use the
new sadl repository that Andy created.

Please review for any misunderstandings or problems.

tuxji committed:
Add CI workflow to build SADL and WebSADL

Set up a GitHub Actions workflow to build SADL source, build WebSADL
extension, publish WebSADL npm package, and push WebSADL image to
Docker Hub.  The workflow (which came from Typefox) has not been
tested and the authentication for the publish npm package step may
require action by Typefox.  I have created the necessary
NPM_AUTH_TOKEN, DOCKERHUB_USERNAME, and DOCKERHUB_TOKEN secrets in the
sadl GitHub settings.

It was not clear to me whether we should push the WebSADL Docker image
to the theiaide or sadl repository in Docker Hub.  I've updated some
documentation files and the workflow on the assumption we'll use the
new sadl repository that Andy created.

Please review for any misunderstandings or problems.

tuxji committed:
Fix sequence not expected YAML error

tuxji committed:
Speed up CI workflow

Rename build job to test and make it only test that macos and windows
can build SADL with Java 11 explicitly specified.  Also cache local
Maven repository to save time.

Make web job build SADL on ubuntu-latest with Java 11, therefore
avoiding need to upload and download language server.  Also cache
local Maven repository to save time too.

tuxji committed:
To observe Maven cache speedup

kittaakos reviewed (commented):
A few remarks:
 - you have changed the Java version from 8 to 11,
 - you have changed the Maven command and you do not use `-U` anymore,
 - I do not understand why we have to run the Maven build twice, GH Actions can share artifacts between jobs, It would be much faster to upload the LS from the Java build, and download it with the web build,
 - I have to know the new Docker and npmjs account names. Otherwise, I cannot add you as a contributor.

Have you had the chance to look into [my workflow](https://github.com/kittaakos/sadl/commit/036ffbc5e6ff22bdf857ddd5d65b205167a11be3) config from my fork?

tuxji committed:
Remove .travis.yml and appveyor.yml

tuxji commented:
To answer your questions, @kittaakos 

- Yes, I have changed the Java version from 8 to 11 because we have had a lot of difficulty using javax.activation.DataSource in SADL on both 8 and 11.  Andy proposed that we move to 11 and no one has objected.
- I changed the Maven command for the following reasons:
   1. GitHub Actions best practice is to cache the local Maven repository and run mvn -B package --file pom.xml.  The difference between install and package is that package will not increase the size of the cached Maven repository by installing newly built jars into the repository.  The difference between verify and package is that verify will check the results of any integration tests.  Errors in unit tests will fail the build before package gets called.  SADL poms do not use maven-failsafe-plugin and therefore do not have any integration tests to be checked by verify.
   2. The mvn clean command is unnecessary because we are building a new checkout of the SADL source.
   3. The --update-snapshots option matters only if SADL uses remote SNAPSHOT dependencies.  I have grep'ed the SADL source and SADL poms do not depend on snapshots of any third party dependencies (none of their versions contain SNAPSHOT, only SADL's own version does).
- There is still only one job (the web job itself) which builds the LS on unix.  I made the test jobs run only on macos and windows, not unix.  The web job runs only on unix.  Therefore, I have saved some time (1 or 2 minutes) because there is no need to make the web job wait for all of the test jobs to finish before it can download the LS and start the web build.  
- Docker account name is `interran` and npmjs account name is `tuxji`.
- Yes, I based the workflow config on your fork's workflow config :).  Looking at it reminded me I hadn't removed .travis.yml and appveyor.yml yet which I've done now, thanks.

kittaakos commented on sadl3/com.ge.research.sadl.parent/com.ge.research.sadl3.documentation/WebSite/sadl3/WebSadlBackend.html:
```diff
@@ -122,9 +122,9 @@ <h3>Pull the Latest WebSADL Image</h3>
-https://hub.docker.com/r/theiaide/sadl/tags</a> for tags.</p>
+<span style="font-family: Courier New">docker pull sadl/sadl:next</span>. 
+See <a href="https://hub.docker.com/r/sadl/sadl/tags">
+https://hub.docker.com/r/sadl/sadl/tags</a> for tags.</p>
```
Well, this won't change. It will remain [`theiaide/sadl`](https://hub.docker.com/r/theiaide/sadl). Only the user changes that we will impersonate on the CI to do the Docker publish.

tuxji replied:
Please clarify whether I should change all of the `sadl/sadl` back to `theiaide/sadl`, @kittaakos 

Andy created a `sadl/` repository on Docker Hub so I assumed he wanted me to publish the `sadl` image there, but I can keep using `theiaide/sadl` if that's what @crapo and you want.

kittaakos commented:
> * I have changed the Java version from 8 to 11

OK.

> * I changed the Maven command for the following reasons:

Good, I am OK with it.

> * There is still only one job

I see two of them: `test` and `web` but it's up to you.

> * Docker account name is `interran` and npmjs account name is `tuxji`.

All set.

tuxji committed:
Rename sadl/sadl back to theiaide/sadl

tuxji commented:
@kittaakos I changed sadl/sadl back to theiaide/sadl and merged the pull request.  The resulting CI failed to publish the npm package and therefore failed to publish the Docker image too.  Will you take a look?  Thanks.

kittaakos commented:
The npm publish step has failed, although the job was green. (Please check the workflow logs):
```
lerna ERR! publish npm ERR! publish Failed PUT 404
lerna ERR! publish npm ERR! code E404
lerna ERR! publish npm ERR! 404 Not found : sadl-extension
lerna ERR! publish npm ERR! 404 
lerna ERR! publish npm ERR! 404  'sadl-extension' is not in the npm registry.
lerna ERR! publish npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
lerna ERR! publish npm ERR! 404 
lerna ERR! publish npm ERR! 404 Note that you can also install from a
lerna ERR! publish npm ERR! 404 tarball, folder, http url, or git url.
lerna ERR! publish 
lerna ERR! publish npm ERR! A complete log of this run can be found in:
lerna ERR! publish npm ERR!     /home/runner/.npm/_logs/2020-11-07T22_25_55_738Z-debug.log
```

The last successful npm publish was a month ago:
```
npm view sadl-extension@next

sadl-extension@0.2.0-next.94f18983 | EPL-1.0 | deps: 8 | versions: 161
An extension for Theia connection language support for the SADL DSL

keywords: theia-extension

dist
.tarball: https://registry.npmjs.org/sadl-extension/-/sadl-extension-0.2.0-next.94f18983.tgz
.shasum: 85d1edcff6c001049aa5955d4dc2c616441e0cd9
.integrity: sha512-NYsgfAnm2QvpHAPZBkRoUh+BhDxM9SSiq2KJ340xty6nv6r7e0I2C7LE0cAh4HO8M0o6MDmujzBKLXqaLra4UA==
.unpackedSize: 121.5 MB

dependencies:
@theia/core: next       @theia/filesystem: next @theia/monaco: next     @theia/output: next
@theia/editor: next     @theia/languages: next  @theia/navigator: next  valid-filename: ^3.1.0

maintainers:
- typefox-team <npm@typefox.io>
- crapo <crapo@research.ge.com>

dist-tags:
latest: 0.1.9              next: 0.2.0-next.94f18983

published a month ago by crapo <crapo@research.ge.com>
```

kittaakos commented:
You can try to add the `ACTIONS_RUNNER_DEBUG` and the `ACTIONS_STEP_DEBUG` secrets to the workflow, both with `true` value, to see more details in the logs.

kittaakos commented:
According to this old comment, the E404 error message can be misleading, and it is actually an authentication issue: https://github.com/npm/npm/issues/8045#issuecomment-103947683

I have invited you as a collaborator to the package? Did you receive it? If yes, please verify your secrets on GH Actions. (The verification can be that you set an incorrect secret start another workflow and check if you still get E404.)

tuxji commented:
@kittaakos I think the E404 error message is really a genuine 404 Not Found message.  When I look at my npmjs.com account page, I see that I have 1 package called sadl-extension:

```
Package 1

sadl-extension
Description
An extension for Theia connection language support for the SADL DSL

Publisher
typefox-team
published 0.1.9 • 2 years ago
```

On the <https://www.npmjs.com/package/sadl-extension/access> page, I see that I am listed as a sadl-access maintainer with write access.  Following your advice, I have created the `ACTIONS_RUNNER_DEBUG` and `ACTIONS_STEP_DEBUG` secrets and run the workflow three more times:

1. Run number 8.  Only change was to add the two new secrets.  Same E404 error message.  Downloaded the logs and examined them.  Found nothing that seemed obviously wrong.
2. Run number 9.  Changed my `NPM_AUTH_TOKEN` secret to a token that belongs to Docker Hub, not npmjs.com.  Same E404 error message.  Found no obvious difference in logs.
3. Run number 10.  Created a second personal access token on npmjs.com (this time, `published` type instead of `automation` type).  Changed my `NPM_AUTH_TOKEN` secret to that new token.  Same E404 error message.  Found no obvious difference in logs.

Therefore, I think we are dealing with a genuine 404 Not Found problem of some kind.  The bug report you linked (<https://github.com/npm/npm/issues/8045>) fixed the original issue so that an authentication error would return 403 Forbidden, not 404 Not Found.

kittaakos commented:
> I think the E404 error message is really a genuine 404 Not Found message. When I look at my npmjs.com account page, I see that I have 1 package called sadl-extension:

No, here is your extension: https://www.npmjs.com/package/sadl-extension/v/0.2.0-next.94f18983

kittaakos commented:
> When I look at my npmjs.com account page, I see that I have 1 package called sadl-extension:

By the way, when you can find it how would this be an E404?

tuxji commented:
I added a couple of commands to the workflow to list and print the npm logs in case they had any more information.  After my most recent run failed to publish the sadl-extension package, there were five npm logs:

```
drwxr-xr-x 2 runner docker  4096 Nov  9 16:21 .
drwxr-xr-x 6 runner docker  4096 Nov  9 16:19 ..
-rw-r--r-- 1 runner docker 35999 Nov  9 16:19 2020-11-09T16_19_55_087Z-debug.log
-rw-r--r-- 1 runner docker 35999 Nov  9 16:20 2020-11-09T16_20_16_625Z-debug.log
-rw-r--r-- 1 runner docker 35999 Nov  9 16:20 2020-11-09T16_20_38_157Z-debug.log
-rw-r--r-- 1 runner docker 35999 Nov  9 16:20 2020-11-09T16_20_58_995Z-debug.log
-rw-r--r-- 1 runner docker 35999 Nov  9 16:21 2020-11-09T16_21_19_734Z-debug.log
```

Each npm log had the same size and what looks like the same contents.  The sha sum this time was c1543d21 (the sha sum of my most recent commit).  The npm log shows the JSON body and HTTP headers that were sent to npmjs.com to try to publish the sadl-extension package.  The response once again was E404:

```
39 error publish Failed PUT 404
40 verbose stack Error: Not found : sadl-extension
```

Please take a look at these npm logs in [run number 14](https://github.com/SemanticApplicationDesignLanguage/sadl/runs/1375207864?check_suite_focus=true) to see if you can find anything wrong.  

I also tried running `npm login` from Ubuntu 20.04 in WSL2 on my own computer.  I entered everything that `npm login` prompted me to enter.  I could not login with `tuxji` and my publish token as the password but I was able to login with `tuxji` and my real password.  I don't know if that is significant or not.

kittaakos commented:
@crapo, can we try the npm publishing with your token?

crapo commented:
I'm fine with you using my token. Do I need to do something? I don't have access to the email any longer where you told me what to do when this was set up. Note also that crapo@research.ge.com, crapo@ge.com are no longer valid email addresses.

kittaakos commented:
> Do I need to do something?

@tuxji can add your npmjs token temporarily to the repository. I do not have access to anything and I do not know how you did set up Travis earlier; my hands are tied.

> Note also that [crapo@research.ge.com](mailto:crapo@research.ge.com), [crapo@ge.com](mailto:crapo@ge.com) are no longer valid email addresses.

Probably it does not matter.

tuxji commented:
@crapo NPM_AUTH_TOKEN is a repository secret which you can update yourself.  It's easy to do - instructions are [here](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets) and you go [here](https://github.com/SemanticApplicationDesignLanguage/sadl/settings/secrets/actions) to update NPM_AUTH_TOKEN's value.  After you've updated the secret, please go [here](https://github.com/SemanticApplicationDesignLanguage/sadl/actions?query=workflow%3A%22SADL+continuous+build%22), click the **Run workflow** button, and when it finishes, please post here to let us know whether it worked or not.

crapo commented:
@tuxji , the workflow failed. I don't see how to observe the cause of failure.

kittaakos commented:
Thank you for checking, @crapo. I will set up a build on my fork with the [`typefox-team`](https://www.npmjs.com/~typefox-team) user.

tuxji commented:
@kittaakos I've given you access to this repo's settings temporarily so you can do your testing right here if that'd save you time.

kittaakos commented:
> I've given you access to this repo's settings temporarily so you can do your testing right here if that'd save you time.

It's all right, thank you so much. I would rather do not want to have access to your secrets. It's all "running" on my fork: and I could reproduce the issue with my token so it's failing. The problem must be somewhere else. I am on it...

tuxji commented:
@kittaakos I saw this article <https://snyk.io/blog/github-actions-to-securely-publish-npm-packages/> and thought it might help us publish the npm package.  I noticed that the article's workflow uses NPM_TOKEN.

kittaakos commented:
> @kittaakos I saw this article https://snyk.io/blog/github-actions-to-securely-publish-npm-packages/ and thought it might help us publish the npm package. I noticed that the article's workflow uses NPM_TOKEN.

Thank you for the pointer, @tuxji. I am currently using [`NODE_AUTH_TOKEN`](https://github.com/actions/setup-node/blob/c6fd00ceb9747fb23ffdf72987450a2664414867/src/authutil.ts#L48) [on my fork](https://github.com/kittaakos/sadl/blob/462e8e020ad363f996a65e2dcc41f38c6a31f850/.github/workflows/build.yml#L95), unfortunately, there are other [issues with the publishing](https://github.com/kittaakos/sadl/runs/1384056410?check_suite_focus=true#step:10:42). I will take care of this CI issue for sure, but I am currently making these changes in my free time; no POs are involved on my side. Please bear with me.

tuxji commented:
Ah, I see part of the problem is adding the token to npm's configuration (the env var's name is not the problem after all).  The article uses this command to add the token - would it help?

`npm config set //registry.npmjs.org/:_authToken ${NPM_TOKEN}`
